futurewei-cloud/arktos
# deploy DNS per network

**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:
flat-network-controller creates API objects listed below (all in the same tenant of the network)  when it detects a new network, besides the kube-dns service:
* service account - coredns, namespace kube-system (only if not existent);
* cluster role - system:coredns (only if not existent);
* cluster role binding - system:codedns (only if not existent);
* configmap - coredns-{network-name}, namespace kube-system;
* deployment - coredns-{network-name}, namespace kube-system.

The deployment specifies 1 replica of coredns pod as it starts.

DNS pod runs coredns binary, will resolve the service name query for the tenant, particularly it sets up the DNS aliases for 
* kube-dns.kube-system.svc.cluster.local. (pointing to kube-dns-{network-name}.kube-system.svc.cluster.local.)
* kubernetes.default.svc.cluster.local. (pointing to kubernetes-{network-name}.default.svc.cluster.local.)

This feature is depicted in [DNS section, multi-tenant network design](https://github.com/futurewei-cloud/arktos/blob/master/docs/design-proposals/multi-tenancy/multi-tenancy-network.md#dns)

**How to verify on local dev env**:
1. start local dev Arktos by ```hack/arktos-up.sh```
2. start flat-network-controller by ```./flat-network-controller --kubeconfig=~/.kube/config```, with the config file having
```yaml
apiVersion: v1
clusters:
- cluster:
    server: http://127.0.0.1:8080
  name: local
contexts:
- context:
    cluster: local
    user: ""
  name: local-ctx
current-context: local-ctx
kind: Config
preferences: {}
users: []
```
3. create the network CRD by ```./cluster/kubectl.sh create -f pkg/controller/artifacts/crd-network.yaml```
4. create network test2 in the system tenant, with following yaml
```yaml
apiVersion: "arktos.futurewei.com/v1"
kind: Network
metadata:
  name: test2
spec:
  type: flat
```
5. verify that coredns-test2 pod is created and running
```
$ ./cluster/kubectl.sh get pod -n kube-system -o wide
coredns-test2-6dbf58b74b-c7mcz    2929372451239929058   1/1     Running   2  10.88.0.127
```
6. verify the DNS server is able to answer service name query properly
```
$ dig  @10.88.0.127 kube-dns-test2.kube-system.svc.cluster.local +short
10.0.0.86
$ dig  @10.88.0.127 kube-dns.kube-system.svc.cluster.local +short
10.0.0.86
```
the ip address returned is that of the service kube-system/kube-dns-test2

**Does this PR introduce a user-facing change?**:
YES
```release-note
1. flat-network-controller accepts argument --cluster-domain to specify the internal domain name of the cluster. By default, it is "cluster.local"
2. each network (of flat type) will has one DNS deployment
```
